url1f1de5fc71_cooijmanstim_hobo_tgz-pJ8-hobo_TestingEnviromentJ8/src/hobo/Belief.java

```diff
@@ -0,0 +1,636 @@
+		double p = 1;
+		CardBag cards = State.INITIAL_DECK.clone();
+		cards.removeAll(s.discarded);
+		cards.removeAll(s.open_deck);
+		Set<Mission> missions = EnumSet.allOf(Mission.class);
+		for (PlayerState ps: s.playerStates()) {
+			if (ps.handle == player) continue;
+			int kmissions = ps.missions.size();
+			if (ps.drawn_missions != null) kmissions += ps.drawn_missions.size();
+			p /= Util.multivariate_hypergeometric(ps.hand.multiplicities(), cards.multiplicities());
+			p /= Util.binomial_coefficient(kmissions, missions.size());
+			cards.removeAll(ps.hand);
+			missions.removeAll(ps.missions);
+			if (ps.drawn_missions != null) missions.removeAll(ps.drawn_missions);
+		}
+		return p;
+	}
+
+	// modifies s
+	public void sampleCards(State s) {
+		CardBag unknown = cards_of_unknown_location.clone();
+
+		CardBag deck = known_deck_cards.clone();
+		deck.addAll(unknown.remove_sample(s.deck.size() - known_deck_cards.size(), random));
+		s.deck = deck;
+
+		for (int i = 0; i < players.length; i++) {
+			if (i == player)
+				continue;
+			PlayerState ps = s.playerState(i);
+			PlayerBelief pb = players[i];
+			CardBag hand = pb.known_cards.clone();
+			hand.addAll(unknown.remove_sample(ps.hand.size() - hand.size(), random));
+			ps.hand = hand;
+		}
+
+		assert(unknown.isEmpty());
+	}
+
+	public double likelihoodOfCards(State s) {
+		double p = 1;
+
+		CardBag unknown = cards_of_unknown_location.clone();
+
+		CardBag unknown_deck_cards = s.deck.clone();
+		assert(unknown_deck_cards.containsAll(known_deck_cards));
+		unknown_deck_cards.removeAll(known_deck_cards);
+
+		p *= unknown.probabilityOfSample(unknown_deck_cards);
+		unknown.removeAll(unknown_deck_cards);
+
+		for (int i = 0; i < players.length; i++) {
+			if (i == player)
+				continue;
+
+			CardBag unknown_hand_cards = s.playerState(i).hand.clone();
+			assert(unknown_hand_cards.containsAll(players[i].known_cards));
+			unknown_hand_cards.removeAll(players[i].known_cards);
+
+			p *= unknown.probabilityOfSample(unknown_hand_cards);
+			unknown.removeAll(unknown_hand_cards);
+		}
+		
+		return p;
+	}
+	
+	// modifies s
+	public void sampleMissions(State s) {
+		// get a working copy of this matrix
+		double[][] player_mission_suspicion = Util.clone(this.player_mission_suspicion);
+
+		// determine how many missions we need to sample for each player
+		int[] ns = new int[players.length];
+		for (int i = 0; i < players.length; i++) {
+			PlayerState ps = s.playerState(i);
+			ns[i] = ps.missions.size();
+			ps.missions = EnumSet.noneOf(Mission.class);
+		}
+
+		// players' missions will be removed from this as we go along
+		s.missions = EnumSet.allOf(Mission.class);
+
+		// make sure our own drawn_missions aren't assigned to other players
+		Set<Mission> ms = s.playerState(player).drawn_missions;
+		if (ms != null) {
+			for (Mission m: ms) {
+				s.missions.remove(m);
+				for (int k = 0; k < players.length; k++)
+					player_mission_suspicion[m.ordinal()][k] = 0;
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 328:334, 438:444.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.